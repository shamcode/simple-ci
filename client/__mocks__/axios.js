const METHODS = [
    'get',
    'post',
    'put',
    'delete'
];

const FIXTURES = {};
const MOCKS = {};

function mockMethodFactory( method ) {
    return jest.fn().mockImplementation(
        ( url ) => {
            if ( !FIXTURES[ method ].hasOwnProperty( url ) ) {
                throw new Error( `Missing fixture for ${method.toUpperCase()} url = "${url}"` );
            }
            const { success, data } = FIXTURES[ method ][ url ];
            return Promise[ success ? 'resolve' : 'reject' ]( {
                data
            } );
        }
    );
}

METHODS.forEach(
    method => {
        MOCKS[ method ] = mockMethodFactory( method );
        FIXTURES[ method ] = {};
    }
);

const defaultMocksData = ( () => {
    const data = {};
    data.chain = {
        id: 2,
        name: 'Test chain',
        command: 'echo 1'
    };
    data.project = {
        id: 1,
        name: 'Test',
        cwd: '/tmp/',
        chains: [ data.chain ]
    };
    return data;
} )();

export default {
    defaultMocksData: defaultMocksData,
    mocks: MOCKS,
    create: jest.fn().mockImplementation( () => {
        return {
            ...MOCKS,
            interceptors: {
                request: {
                    use: () => {}
                },
                response: {
                    use: () => {}
                }
            }
        };
    } ),
    use( method, url, data, success = true ) {
        if ( !METHODS.includes( method ) ) {
            throw new Error( `Unknown method: "${method}"`);
        }
        FIXTURES[ method ][ url ] = {
            data,
            success
        };
        return this;
    },
    useDefaultMocks() {
        this
            .use( 'get', '/projects', [ defaultMocksData.project ] )
            .use( 'get', '/projects/1', defaultMocksData.project )
            .use( 'get',  '/chains/2', defaultMocksData.chain );
        return this;
    }
};