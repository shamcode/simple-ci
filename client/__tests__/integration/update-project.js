import setup, { app } from './helpers'
import axios from 'axios';
jest.mock( 'axios' );

beforeEach( () => {
    jest.resetModules();
    jest.clearAllMocks();
    setup();
} );

it( 'update project', async() => {
    expect.assertions( 7);

    axios
        .useDefaultMocks()
        .use( 'put', '/projects/1', null );

    await app.start();
    await app.project.open();

    await app.click( '[href="projects/1/edit"]' );
    app.checkBody();
    await app.waitRendering();
    app.checkBody();
    expect( document.querySelector( '[name="name"]' ).value ).toBe( axios.defaultMocksData.project.name );
    expect( document.querySelector( '[name="cwd"]' ).value ).toBe( axios.defaultMocksData.project.cwd );

    const formData = {
        name: 'Test name',
        cwd: 'test cwd'
    };
    app.form.fill( 'name', formData.name );
    app.form.fill( 'cwd', formData.cwd );
    await app.form.submit();
    expect( axios.mocks.put ).toHaveBeenCalledTimes( 1 );
    expect( axios.mocks.put.mock.calls[ 0 ][ 1 ] ).toEqual( formData );

    app.checkBody();
} );

