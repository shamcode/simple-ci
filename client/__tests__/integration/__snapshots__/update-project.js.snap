// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update project 1`] = `
"<header>
  <a href=\\"\\" class=\\"logo\\">Simple CI</a>
</header>
<!--Header-->
<main>
  <div class=\\"page-data-loading\\">
    <div class=\\"sk-cube-grid\\">
      <div class=\\"sk-cube sk-cube1\\"></div>
      <div class=\\"sk-cube sk-cube2\\"></div>
      <div class=\\"sk-cube sk-cube3\\"></div>
      <div class=\\"sk-cube sk-cube4\\"></div>
      <div class=\\"sk-cube sk-cube5\\"></div>
      <div class=\\"sk-cube sk-cube6\\"></div>
      <div class=\\"sk-cube sk-cube7\\"></div>
      <div class=\\"sk-cube sk-cube8\\"></div>
      <div class=\\"sk-cube sk-cube9\\"></div>
    </div>
  </div>
  <!--PageDataLoading-->
  <!--if-->
</main>
<footer> Sham 2018 </footer>
<!--Footer-->"
`;

exports[`update project 2`] = `
"<header>
  <a href=\\"\\" class=\\"logo\\">Simple CI</a>
</header>
<!--Header-->
<main>
  <!--if-->
  <!--ErrorsPanel-->
  <div class=\\"project-create-form\\">
    <form>
      <div class=\\"group\\">
        <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
        <label>Name</label>
      </div>
      <div class=\\"group\\">
        <input name=\\"cwd\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
        <label>Current Work Directory</label>
      </div>
      <div class=\\"group\\">
        <button type=\\"submit\\"><span>Update</span></button>
      </div>
    </form>
  </div>
  <!--ProjectForm-->
  <!--if-->
</main>
<footer> Sham 2018 </footer>
<!--Footer-->"
`;

exports[`update project 3`] = `
"<header>
  <a href=\\"\\" class=\\"logo\\">Simple CI</a>
</header>
<!--Header-->
<main>
  <!--if-->
  <!--ErrorsPanel-->
  <div class=\\"project-info\\">
    <h3>Test</h3>
    <dl>
      <dt>/tmp/</dt>
      <dd>Current work directory.</dd>
      <dd>This directory will be used when running tasks.</dd>
      <dt>
        <a href=\\"projects/1/edit\\" class=\\"\\">Edit</a>
      </dt>
      <dd>Edit project name &amp; cwd</dd>
      <dt>
        <a href=\\"projects/1/delete\\" class=\\"project-delete-link\\">Delete</a>
      </dt>
      <dd>Delete project</dd>
    </dl>
  </div>
  <!--Info-->
  <!--if-->
</main>
<footer> Sham 2018 </footer>
<!--Footer-->"
`;
